cocos/tween/tween-progress.ts

```diff
@@ -0,0 +1,92 @@
+ of this software and associated documentation files (the "Software"), to deal
+ in the Software without restriction, including without limitation the rights to
+ use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
+ of the Software, and to permit persons to whom the Software is furnished to do so,
+ subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in
+ all copies or substantial portions of the Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+*/
+
+import type { ITweenCustomProperty, ITweenCustomPropertyStartParameter } from './tween';
+import { Spline, SplineMode } from '../core/geometry';
+import { Vec3 } from '../core/math';
+
+function createSplineProperty (mode: SplineMode, knots: ReadonlyArray<Vec3>): ITweenCustomProperty<Vec3> {
+    let spline: Spline | null = null;
+    return {
+        value: knots.length > 0 ? knots[knots.length - 1] : Vec3.ZERO,
+        progress (start: Vec3, end: Vec3, current: Vec3, ratio: number): Vec3 {
+            return spline!.getPoint(ratio);
+        },
+        clone (v: Vec3): Vec3 {
+            return Vec3.clone(v);
+        },
+        add (a, b): Vec3 {
+            return a.clone().add(b);
+        },
+        sub (a, b): Vec3 {
+            return a.clone().subtract(b);
+        },
+        onStart (param: ITweenCustomPropertyStartParameter<Vec3>): void {
+            const { start, end, relative, reversed } = param;
+            spline = Spline.create(mode);
+            spline.addKnot(start);
+            let reversedLast: Vec3 | null = null;
+            if (relative && reversed) {
+                reversedLast = new Vec3();
+                Vec3.subtract(reversedLast, start, knots[knots.length - 1]);
+            }
+            for (let i = 0, len = knots.length; i < len; ++i) {
+                const v = reversed ? knots[len - 1 - i] : knots[i];
+                if (relative) {
+                    if (reversed) {
+                        // Skip the start point ( i = 0 )
+                        if (i > 0) {
+                            spline.addKnot(reversedLast!.clone().add(v));
+                        }
+                    } else {
+                        spline.addKnot(start.clone().add(v));
+                    }
+                } else {
+                    spline.addKnot(v);
+                }
+            }
+
+            if (relative && reversed) {
+                spline.addKnot(end);
```

Comment:
Why only need to add `end` when `elative && reversed`?

Reply:
Suppose that we have 5 knots, like [1, 2, 3, 4, 5], when we run `tween.to`, the path will be
( target's current position, 0 ) -> 1 -> 2 -> 3 -> 4 -> 5,  and 5 is the end knot. So we get 6 knots for the curve.
In `reversed` mode, the sequence will be 
```
(target's current position, 5 ) -> 4 -> 3 -> 2 -> 1 -> 0
```

, and 0 is the end knot in reverse mode, 5 needs to be skipped since we have already added the `start` knot.
There is not 0 knot in the original array developer passed in, so we need to add the `0` knot which is `end`.